### ISSUE TITLE
“self signed certificate” error during extension load

### ISSUE THREAD

**Issue by strogonoff** [paneron/paneron#39]
Interestingly, extension code is downloaded anyway.

If this is a network configuration issue, this may be causing other problems (such as failing to clone repositories), but so far the only confirmed error is the self signed certificate error when accessing `registry.npmjs.com`.

My theory is that employer forces a self-signed certificate, and this messes up with secure HTTPS requests. (Trying to confirm this.)

Attaching log from running Paneron on Windows:
![error1](https://user-images.githubusercontent.com/292959/141110510-ed44bb73-a9f8-4acd-805a-f0663b5eb770.png)

cc @wkwong-ribose

**Comment by strogonoff** [comment]
A way to check for an extra certificate is to open a package URL (e.g., `https://registry.npmjs.org/left-pad`) in normal browser, and use browser functionality to see the certificate chain.

It should normally look like this (no self-signed certificates in chain):
<img width="551" alt="Screenshot 2021-11-10 at 12 57 38 PM" src="https://user-images.githubusercontent.com/292959/141110865-eaa44c6f-54b8-4fe5-98af-6779db39d8ca.png">

**Comment by strogonoff** [comment]
Attempting to address in https://github.com/paneron/paneron/commit/c5c78d5cafb64dec0772102672065b08774e70c6, will test with Maryam this Monday

**Comment by odeclas** [comment]
> A way to check for an extra certificate is to open a package URL (e.g., `https://registry.npmjs.org/left-pad`) in normal browser, and use browser functionality to see the certificate chain.

If it helps at all we're not seeing any issues in Certificate handling for the reference site, either on VPN or without VPN. Tested Chrome and Edge in case there were differences. 

> It should normally look like this (no self-signed certificates in chain): <img alt="Screenshot 2021-11-10 at 12 57 38 PM" width="551" src="https://user-images.githubusercontent.com/292959/141110865-eaa44c6f-54b8-4fe5-98af-6779db39d8ca.png">

We do get different routing, but I guess that's to be expected.
![image](https://user-images.githubusercontent.com/384986/141767209-0edf3b06-bd4c-4b31-a437-591b2ebf5fd3.png)

**Comment by strogonoff** [comment]
@odeclas I believe different mechanisms of SSL handling may be used. In particular, I suspect Node bundles its own trusted root CAs, and uses that by default instead of trusting user/system root CAs.

I saw this happen before (some project relied on an SSL library used its own bundled set of root CAs rather than the system, and so got SSL errors when making requests even though normal `curl` on the same machine worked fine).

So far I suspect there are two issues, as identified after Thursday’s test with Maryam:

- The error with extension metadata retrieval. I believe the plugin library we use might be handling SSL differently than the system when making requests to `registry.npmjs.com`, with compatibility issues on some work machines.
- Our “local extension” loading mechanism, apparently, still made network requests to fetch extension metadata (resulting in the above error), which it shouldn’t do if extension is being loaded from local filesystem.

**Comment by strogonoff** [comment]
Update:

* *The problem with self signed certificate error:*
  * Here is an [upstream feature request](https://github.com/nodejs/node/issues/3159#issuecomment-658508977) that summarizes the problem. In other words, you don’t get any errors in a regular browser because it’s built with access to relevant OS-specific custom certificate sources, but our application platform doesn’t have that access.
  * We could dynamically provide the “special” cert at runtime, but even if we allow the user to specify the cert it wouldn’t help because self-signed certificates installed by IT are typically not made available to the user.
  * We *can* address this with some cooperation of each respective user organization’s IT team. This implies creation of a separate Paneron version for each organization where employee machines use self-signed root certificate.
  * We could avoid this if we avoid using Node’s cross-platform `fetch()` API. It could be possible once we [change extension mechanism](https://github.com/paneron/paneron/issues/36), which would also take us much closer to Paneron Web.
  * We wouldn‘t have this issue with Paneron Web that runs from user’s browser (and therefore uses).

* *The problem with local extensions failing because of network error:*
  * Regular extensions don’t require network when locally installed, and shouldn’t cause the error.
  * Extensions that have extra dependencies, however, will result in the same problem as above, because extra dependencies are _not_ locally installed and will need to be downloaded.
  * _HLS extension started to bundle Glossarist recently, and started causing this error._
  * This can be addressed by packaging extension dependencies together with the extension.

**Comment by odeclas** [comment]
@strogonoff  I'm just about following the thread, but the bit that I don't understand is that it is fetching metadata from the repository successfully, so is there something different about fetching the dataset files:
![image](https://user-images.githubusercontent.com/384986/141825068-e07448ee-b381-400f-b385-92c565df9ef4.png)

**Comment by strogonoff** [comment]
Yes, I'm looking into that too. We use different libraries for those two things, and the library used for fetching repository data seems not to have this problem.

It gives me hope that we can do the same with extensions, and it worries me a little because that logic may be vulnerable to potential MITM attacks in a way that extension loading isn’t.

Although running bad extension code is a bigger security issue then working with a bad repo (the former can lead to machine infection/takeover, the latter to data leak and even then not clear how feasibly).
